Update the rest of the registry docs (#2169)
viamrobotics/docs: docs/registry/create/_index.md

  ]
description: "Use the Viam module system to implement modular resources that can be included in any Viam-powered smart machine."
aliases:
  - "/extend/registry/create/"
  - "/modular-resources/create/"
no_list: true
---

Viam provides many built-in models. If you cannot find existing modular resources from the [Viam Registry](/registry/#the-viam-registry) that support your hardware or software, you can add new {{< glossary_tooltip term_id="module" text="modules" >}}.

A _module_ provides one or more {{< glossary_tooltip term_id="modular-resource" text="modular resources" >}}, which add {{< glossary_tooltip term_id="resource" text="resource" >}} {{< glossary_tooltip term_id="type" text="types" >}} or {{< glossary_tooltip term_id="model" text="models" >}} that are not built into Viam.
Modules run alongside `viam-server` as separate processs, communicating with `viam-server` over UNIX sockets.

> When the module initializes, it registers those pairs on your robot, making the functionality defined by that pair available for use.

Author: ```suggestion
When a module initializes, it registers its {{< glossary_tooltip term_id="model" text="model or models" >}} and the associated [APIs](/program/apis) with `viam-server`, making the new model available for use.
```

description: "Use the Viam module system to implement modular resources that can be included in any Viam-powered smart machine."
aliases:
  - "/extend/modular-resources/create/"
  - "/modular-resources/create/"
no_list: true
---

> Viam provides many built-in models. If you cannot find existing modular resources from the [Viam Registry](/registry/#the-viam-registry) that support your hardware or software, you can add new {{< glossary_tooltip term_id="module" text="modules" >}}.

Author: ```suggestion
Viam provides built-in support for a variety of different [components](/components) and [services](/services/), but you can also add support for unsupported resources by creating a module.
A _module_ provides one or more {{< glossary_tooltip term_id="modular-resource" text="modular resources" >}}, which add support for {{< glossary_tooltip term_id="resource" text="resource" >}} {{< glossary_tooltip term_id="type" text="types" >}} or {{< glossary_tooltip term_id="model" text="models" >}} that are not built into Viam.

You can browse existing modules in the [Viam Registry](/registry/#the-viam-registry) to find one that supports your custom hardware or software, or you can write your own module.
```

description: "Use the Viam module system to implement modular resources that can be included in any Viam-powered smart machine."
aliases:
  - "/extend/modular-resources/create/"
  - "/modular-resources/create/"
no_list: true
---

Viam provides built-in support for a variety of different [components](/components) and [services](/services/), but you can also add support for unsupported resources by creating a module.
A _module_ provides one or more {{< glossary_tooltip term_id="modular-resource" text="modular resources" >}}, which add support for {{< glossary_tooltip term_id="resource" text="resource" >}} {{< glossary_tooltip term_id="type" text="types" >}} or {{< glossary_tooltip term_id="model" text="models" >}} that are not built into Viam.

You can browse existing modules in the [Viam Registry](/registry/#the-viam-registry) to find one that supports your custom hardware or software, or you can write your own module.

> A _module_ provides one or more {{< glossary_tooltip term_id="modular-resource" text="modular resources" >}}, which add {{< glossary_tooltip term_id="resource" text="resource" >}} {{< glossary_tooltip term_id="type" text="types" >}} or {{< glossary_tooltip term_id="model" text="models" >}} that are not built into Viam.

Author: ```suggestion
```

1. [Code a new resource model](#code-a-new-resource-model) server.
2. [Code a main program](#code-a-main-entry-point-program).
3. [Compile or package](#compile-the-module-into-an-executable) the module into a single executable.

### Code a new resource model

{{% alert title="Naming your model" color="tip" %}}

Use the naming schema: `namespace:repo-name:name`.

For more information see [Naming your model](/registry/upload/#naming-your-model-namespacerepo-namename).

{{% /alert %}}

{{< tabs >}}
{{% tab name="Python" %}}
Your new resource model server must have all the methods that the Viam RDK requires, and should match the built-in API client {{< glossary_tooltip term_id="subtype" text="subtype" >}} like ['rdk:component:base'](https://python.viam.dev/autoapi/viam/components/base/index.html).

Create a folder for your module and save your code as a file named <file>my_modular_resource.py</file> inside.

The following example module registers a modular resource implementing Viam's built-in [Base API](/components/base/#api) [(rdk:service:base)](/registry/key-concepts/#models) as a new model, `"mybase"`, using the model family `acme:demo:mybase`.

> <file>my_base.py</file> implements "mybase", a custom model of the base component.

Reviewer: Opt: could remove this line now that we have line 133 above, which does it better.

1. [Code a new resource model](#code-a-new-resource-model) server.
2. [Code a main program](#code-a-main-entry-point-program).
3. [Compile or package](#compile-the-module-into-an-executable) the module into a single executable.

### Code a new resource model

{{% alert title="Naming your model" color="tip" %}}

Use the naming schema: `namespace:repo-name:name`.

For more information see [Naming your model](/registry/upload/#naming-your-model-namespacerepo-namename).

{{% /alert %}}

{{< tabs >}}
{{% tab name="Python" %}}
Your new resource model server must have all the methods that the Viam RDK requires, and should match the built-in API client {{< glossary_tooltip term_id="subtype" text="subtype" >}} like [`rdk:component:base`](https://python.viam.dev/autoapi/viam/components/base/index.html).

Create a folder for your module and save your code as a file named <file>my_modular_resource.py</file> inside.

The following example module registers a modular resource implementing Viam's built-in [Base API](/components/base/#api) [(rdk:service:base)](/registry/key-concepts/#models) as a new model, `"mybase"`, using the model family `acme:demo:mybase`.

> <file>my_base.py</file> implements "mybase", a custom model of the base component.

Author: ```suggestion
```

            self.right.is_powered(extra=extra, timeout=timeout)[0]
```

</details>
<br>

Additional example modules are available in the [Python SDK GitHub repository](https://github.com/viamrobotics/viam-python-sdk/tree/main/examples).

{{% /tab %}}
{{% tab name="Go"%}}

> Your new resource model server must have all the methods that the Viam RDK requires, and should match the built-in API client {{< glossary_tooltip term_id="subtype" text="subtype" >}} like ['rdk:component:base'](https://pkg.go.dev/go.viam.com/rdk/components/base#pkg-functions).

Reviewer: ```suggestion
Your new resource model server must have all the methods that the Viam RDK requires, and should match the built-in API client {{< glossary_tooltip term_id="subtype" text="subtype" >}} like [`rdk:component:base`](https://pkg.go.dev/go.viam.com/rdk/components/base#pkg-functions).
```

Create a folder for your module and save your code as a file named <file>my_modular_resource.go</file> inside.

The following example module registers a modular resource implementing Viam's built-in [Base API](/components/base/#api) [(rdk:service:base)](/registry/key-concepts/#models) as a new model, `"mybase"`, using the model family `acme:demo:mybase`.

> <file>mybase.go</file> implements "mybase", a custom model of the base component, and registers the new model and API helper functions with the Go SDK.

Reviewer: Opt: could remove this line as well in favor of line 316, though there's a but more usefulness to go's version here.

            self.right.is_powered(extra=extra, timeout=timeout)[0]
```

</details>
<br>

Additional example modules are available in the [Python SDK GitHub repository](https://github.com/viamrobotics/viam-python-sdk/tree/main/examples).

{{% /tab %}}
{{% tab name="Go"%}}

Your new resource model server must have all the methods that the Viam RDK requires, and should match the built-in API client {{< glossary_tooltip term_id="subtype" text="subtype" >}} like [`rdk:component:base`](https://pkg.go.dev/go.viam.com/rdk/components/base#pkg-functions).

Create a folder for your module and save your code as a file named <file>my_modular_resource.go</file> inside.

The following example module registers a modular resource implementing Viam's built-in [Base API](/components/base/#api) [(rdk:service:base)](/registry/key-concepts/#models) as a new model, `"mybase"`, using the model family `acme:demo:mybase`.

> <file>mybase.go</file> implements "mybase", a custom model of the base component, and registers the new model and API helper functions with the Go SDK.

Author: ```suggestion
```

```

</details>
<br>

> The code for the custom model (<file>mybase.go</file>) and module entry point file (<file>main.go</file>) is adapted from the full demo modules available in the [in the RDK GitHub repository](https://github.com/viamrobotics/rdk/blob/main/examples/customresources).

Reviewer: supernit: link title text to GH.

```suggestion
The code for the custom model (<file>mybase.go</file>) and module entry point file (<file>main.go</file>) is adapted from the full demo modules available in the [in the RDK](https://github.com/viamrobotics/rdk/blob/main/examples/customresources).
```

> Additional examples are available in the [in the RDK GitHub repository](https://github.com/viamrobotics/rdk/blob/main/examples/).

Reviewer: This links looks identical to previous sentence, but is for more examples. This title text change helps clarify that:

```suggestion
Additional examples are available in the [examples directory of the RDK](https://github.com/viamrobotics/rdk/blob/main/examples/).
```

{{< tabs name="Sample SDK Main Program Code">}}
{{% tab name="Python"%}}

The main program starts the module.
<file>main.py</file> is the module's entry point file.

> Import your custom model and API into the main program and register them with your chosen SDK.

Reviewer: In python context (tab) so we can assume python:

```suggestion
Import your custom model and API into the main program and register them with the Python SDK.
```

</details>

{{% alert title="Important" color="note" %}}

You must define all functions belonging to a built-in resource subtype's API if defining a new model.
Otherwise, the class will not instantiate.

If you do not wish to implement all methods, raise an`NotImplementedError()` in the body of functions you do not want to implement or put `pass`.

Additionally, return any values designated in the function's return signature, typed correctly.

{{% /alert %}}

{{% /tab %}}
{{% tab name="Go"%}}

The main program starts the module.
<file>main.go</file> is the module's entry point file.

> Import your custom model and API into the main program and register them with your chosen SDK.

Reviewer: Nit: in go context, so we can assume go. But is RDK correct here:

```suggestion
Import your custom model and API into the main program and register them with the RDK.
```

</details>

{{% alert title="Important" color="note" %}}

You must define all functions belonging to a built-in resource subtype's API if defining a new model.
Otherwise, the class will not instantiate.

If you do not wish to implement all methods, raise `errUnimplemented` for the functions you do not want to implement.

Additionally, return any values designated in the function's return signature, typed correctly.

{{% /alert %}}

{{% /tab %}}
{{< /tabs >}}

### Compile the module into an executable

To [add a module](/registry/configure/) to the configuration of your robot, you need to have an [executable](https://en.wikipedia.org/wiki/Executable) that:

> - runs your module when executed,
> - can take a local socket as a command line argument,
> - and cleanly exits when sent a termination signal.

Reviewer: I love this approach to complicated, multi-part sentences! Tiny nits:

```suggestion
- runs your module when executed,
- takes a local socket as a command line argument, and
- cleanly exits when sent a termination signal.
```

Your options for completing this step are flexible, as this file does not need to be in raw binary format.

{{% tabs %}}
{{% tab name="Python" %}}

One option is to create and save a new shell script (<file>.sh</file>) that runs your module at your entry point (main program) file.

> Make sure to set up a Python virtual environment in the directory your module is in to ensure your module has access to any required libraries:

Reviewer: Alternatively, we could just link to this resource:

https://docs.viam.com/program/python-venv/

Author: that doesn't mention the `exec $PYTHON <your-src-dir-if-inside>/main.py $@` part

   ```sh { class="command-line" data-prompt="$"}
   viam-sdk
   ```

   Add additional dependencies as needed.
   See the [pip `requirements.txt` file documentation](https://pip.pypa.io/en/stable/reference/requirements-file-format/) for more information.

2. Add a shell script that creates a new virtual environment, installs the dependencies listed in `requirements.txt`, and runs the module entry point file `main.py`:

   ```sh { class="command-line" data-prompt="$"}
    #!/bin/sh
    cd `dirname $0`

    # Create a virtual environment to run our code
    VENV_NAME="venv"
    PYTHON="$VENV_NAME/bin/python"

    python3 -m venv $VENV_NAME
    $PYTHON -m pip install -r requirements.txt -U # remove -U if viam-sdk should not be upgraded whenever possible

    # Be sure to use `exec` so that termination signals reach the python process,
    # or handle forwarding termination signals manually
    exec $PYTHON <your-src-dir-if-inside>/main.py $@
   ```

   To make your shell script executable, run the following command in your terminal:

   ```sh { class="command-line" data-prompt="$"}
   sudo chmod +x <your-file-path-to>/<run.sh>
   ```

> See the [Python virtual environment documentation](https://docs.python-guide.org/dev/virtualenvs/) for more information.

Author: ```suggestion
See [Prepare your Python Virtual Environment](/program/python-venv/) for more information.
```

> See the [Python virtual environment documentation](https://docs.python-guide.org/dev/virtualenvs/) for more information.

Reviewer: If we retain parallel instructions, recommend changing this See Also to point to https://docs.viam.com/program/python-venv/ instead.

{{% /tab %}}
{{< /tabs >}}

The examples from [Code a new resource model](#code-a-new-resource-model) include this logging setup.

## Next steps

> Once you have created your modular resource, you can use the [Viam CLI](/manage/cli/) to [upload your modular resource](/registry/upload/) to the [Viam registry](https://app.viam.com/registry) to share it with other Viam users or just to other users in your organization.

Reviewer: nit: should these all be `module`?

```suggestion
Once you have created your module, you can use the [Viam CLI](/manage/cli/) to [upload your module](/registry/upload/) to the [Viam registry](https://app.viam.com/registry) to share it with other Viam users or just to other users in your organization.
```
